src/enumerable.cr

```diff
@@ -984,6 +1019,43 @@ module Enumerable(T)
     max_by? &.itself
   end
 
+  # Returns an array of the maximum `k` elements, sorted descending.
+  #
+  # It compares using `<=>` so it will work for any type that supports that method.
+  #
+  # ```
+  # [7, 5, 2, 4, 9].max(3) # => [9, 7, 5]
+  # ["Eve", "Alice", "Bob", "Mallory", "Carol"].max(2) # => ["Mallory", "Eve"]
+  # ```
+  #
+  # Raises `Enumerable::ArgumentError` if `k` is negative or if any elements are
+  # not comparable.
+  def max(k : Int) : Array(T)
+    raise ArgumentError.new("negative size #{k}") if k < 0
+    data = to_a
```

Comment:
If the receiver is already an `Array`, this `to_a` simply returns `self` and so the method would mutate the receiver when it shouldn't. Ditto for the other overloads.

Reply:
I wonder if we shouldn't let `Array#to_a` a synonym of `#dup` and always duplicate the array, and let optimizations do the opposite if needed. This is a common pitfall.

Reply:
In an [ideal world](https://github.com/crystal-lang/crystal/pull/12937) one would simply write `[*a]` to get this behavior.

Reply:
Yes, there should definitely be a generic solution to this problem. There are effectively two different semantics that are expected of `#to_a`.
This is completely unrelated to this PR, so let's continue this discussion in #13059

Reply:
This is a common pitfall for these Enumerable and Indexable algorithms. I doubt it's useful to have a public API. There's no such API in Ruby. I wouldn't spend too much time on this

Reply:
Ruby also most certainly had `[*exp]` before Crystal was born 😄 (Ruby then calls `exp.to_a` _before_ doing the actual splat expansion; in particular, `nil.to_a` returns an empty array. So it is relatively important to them that `Array#to_a` doesn't return a duplicate)

Also the reference Ruby implementation's `Enumerable` algorithms are written in C, not Ruby itself, so they never have to worry about the standard library not being utilized by itself when so much happens at the C level.